File: Shashank Shukla 18-CSE/Day-8.java
/*
approach: used iterator to iterate through the main list, if it encounters an element then it is directly added to the result list,
          else recursive call is made to iterate through the nested list. 
*/
public class NestedIterator implements Iterator<Integer> {

    private List<Integer> resList = null;
    private Iterator<Integer> i;
    public NestedIterator(List<NestedInteger> nestedList) {
        resList = new ArrayList<>();
        flat(nestedList);
        i = resList.iterator();
    }
    
    private void flat(List<NestedInteger> nestedList){
        for(NestedInteger curr: nestedList){
            if(curr.isInteger()){
                resList.add(curr.getInteger());
            } else {
                flat(curr.getList());
            }
        }
    }

    @Override
    public Integer next() {
        if(i.hasNext())
            return i.next();
        else 
            return null;
    }

    @Override
    public boolean hasNext() {
        return i.hasNext();
    }
}

File: Aman Dwivedi 18-ECE/Day 8.cpp
https://leetcode.com/problems/flatten-nested-list-iterator/
Tine: O(N)
Space: O(N)

class NestedIterator {
    
    int curr=0; //for current position of original list
    vector<int> ansList; //for storing integers
public:
    
    void solve(vector<NestedInteger> &nestedList) //helper function
    {
        int n=nestedList.size();
        int ele;
        vector<NestedInteger> tempList; //for storing list
        
        //while iterating next element, it can be a integer or a integer list,
        //if it is a integer we take that value and push it to the ansList vector, else if a list of integers, we recursively call the helper function,
        //then the magic of recursion happens it solves that part individually. :p
        for(int i=0;i<n;i++)
        {
            if(nestedList[i].isInteger())
            {
                int ele=nestedList[i].getInteger();
                ansList.push_back(ele);   
            }
            else
            {
                tempList=nestedList[i].getList();
                solve(tempList); //recursion magic 
            }
        }
    }
    NestedIterator(vector<NestedInteger> &nestedList) {
        solve(nestedList); 
    }
    
    int next() {
        int ans=ansList[curr]; //simply returns the next element
        curr++;
        return ans;
    }
    
    bool hasNext() { //check if next element is present or not
        if(curr==ansList.size())
            return false;
        else
            return true;
    }
};
